enhancements/machine-api/maintenance-lock.md

```diff
@@ -0,0 +1,254 @@
+drain of nodes, as well as to construct additional commands to determine
+progress, wait for completion, triage failures, and unwind the process once the
+maintenance activity is completed.
+
+## Motivation
+
+Badly timed updates can make triage an even longer process, but poorly timed
+power events could corrupt the machine and represent a health and safety risk.
+
+Without first-class support from the server, the manual steps needed to perform
+equivalent functionality may lack consistency between admins, present
+opportunities for mistakes, and are not reflected in the UI/consoles for other
+parties to base informed decisions.
+
+All these aspects are especially important in Telco environments where there
+are multiple levels of admins, across different physical locations, managing
+thousands of clusters.
+
+In such an environment, the admin needs to find identify the cluster a
+problematic node belongs to as well as locate and use the corresponding
+`kube.config` and/or pass `oc` the correct cluster name for every call.  Not
+impossible but certainly error prone.
+
+### Goals
+
+- A mechanism for signaling that a node is in a maintenance state
+
+- A metal3 specific operator that performs relevant actions (eg. cordon +
+  drain) when maintenance is requested
+
+- Changes to existing components to disable or delay functions that would
+  affect the node, such as:
+  - Delay upgrades
+  - Disable health checks
+
+- Expose the ability to display and control the maintenance state via the UI
+
+- Expose the ability to query and control the maintenance state programatically
+  via a console
+
+
+### Non-Goals
+
+- To agree on a single set of actions that should be performed across all
+  platforms.
+
+## Proposal
+
+Utilize the existing `Lease` built-in API in the API group
+`coordination.k8s.io` to define a new per-node `Lease` object with Name equal
+to Node name in a newly created dedicated namespace. [Further
+details](https://github.com/kubernetes/enhancements/pull/1411)
+
+Components performing a invasive and/or destructive actions much aquire the
+per-node lease in advance to avoid software updates, reboots, or hardware power
+events while another component is performing a similar action.
+
+Invasive actions include anything that modifes the node's operating system or
+it's configuration (eg. MCO/MCD).  Destructive actions include anything that
+result in the node being deleted (autoscaling, machine healthcheck) or change
+power state.
+
+On metal3 based deployments, a `lifecycle.openshift.io/maintenance` node
+annotation will be created for UIs and consoles to use to put a node into
+maintenance.
+
+Additionally, we define a new Metal3 Maintenance Operator that:
+
+- watches for the `lifecycle.openshift.io/maintenance` annotation to be added
+- acquires the per-node lease
+- cordons the node
+- uses the drain library to move workloads elsewhere
+- reboots the machine at the end of the maintenance window
```

Comment:
Thanks for the feedback @JoelSpeed 

I'm not wedded to the idea of rebooting after the maintenance window ends, but it seemed consistent with OCP's philosophy of wanting things to be "clean".  Especially after an admin had been doing "things" to the node.

Rebooting immediately after would also make it clearer to all parties that whatever happened in the maintenance window is the likely cause should the node not come back healthy.

I wouldn't claim the the flow as written is optimal, but I wouldn't go so far as calling it unsafe. Nodes may disappear without warning at any time.

But you do make a valid point, we should uncordon after the node is rebooted, not prior.

Reply:
> I'm not wedded to the idea of rebooting after the maintenance window ends, but it seemed consistent with OCP's philosophy of wanting things to be "clean". Especially after an admin had been doing "things" to the node.

I can definitely see the value in trying to keep the machines clean, I was just wondering whether this might be a surprising behaviour to some admins? Perhaps it could be opt in/out? I initially thought that the decision as to when/if reboots occur should be left to the person maintaining the machine, but I can see value in enforcing the best practice of rebooting for a clean slate (assuming rebooting does clean the slate?)

> Rebooting immediately after would also make it clearer to all parties that whatever happened in the maintenance window is the likely cause should the node not come back healthy.

That's a nice benefit of rebooting at the end of the window, definitely makes me lean towards opt out if it were an opt in/out scenario

> I wouldn't claim the the flow as written is optimal, but I wouldn't go so far as calling it unsafe. Nodes may disappear without warning at any time.

Yeah, apologies, unsafe is a bit strong, totally agree, it's definitely no worse than a node just disappearing!